README.MD
<img src="https://cdn.jsdelivr.net/gh/czy0729/bangumi@master/preview/bangumi.png" width="100%" />

<p align="center">
  <img alt="Version: 4.7.0" src="https://img.shields.io/badge/Version-4.7.0-0366d6" />
  <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-33CC12" />
  <img alt="Codacy" src="https://api.codacy.com/project/badge/Grade/473bdb1186484703b4069148730a3c12?isInternal=true" />
  <img alt="jsdelivr" src="https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi/badge" />
</p>

---

一个基于`React-Native`的 [Bangumi](https://bgm.tv) 第三方客户端

项目对标 web, 本人并没有原生开发经验, 用于练手 RN, 安卓和 iOS 都做了兼容, 并最大限度保持一致

目前已开发不少于 70 个页面, 几乎涵盖源网站所有页面

Bangumi 是用于管理追番进度（不限于动漫、音乐、日剧、AVG）, 发现和讨论番组的第三方客户端

本人并没有原生开发经验, 用于练手 RN, 希望通过这个平台数据去学习更多

## Download

[安卓 APK 下载](https://github.com/czy0729/Bangumi/releases) | [酷安市场下载](https://www.coolapk.com/apk/235642)

苹果开发者账号已过期, 期间没能过审, 有缘再送审.
[TestFlight](https://testflight.apple.com/join/pAaJ0MWU) | [Expo 试用方法](https://github.com/czy0729/Bangumi/blob/master/IOS-TEST.MD) | [Expo 主页](https://expo.io/@bgm-ios-test/bangumi-pro)

## Features

[特有功能]

- 大部分资源覆盖到自维护 [CDN](https://github.com/czy0729/Bangumi-Subject), 访问速度++
- 图片使用 [CDN](https://github.com/czy0729/Bangumi-OSS), 有后处理, 比源网站更快、质量更高、流量更小
- 当季番剧精准到单集网络播放源 / 漫画源查询 / 找番剧 / 文库
- [小圣杯](https://bgm.tv/dev/app/1143) (角色交易空气游戏, App 独立设计 UI 和交互)

[基本功能]

- 番组进度管理
- 条目、角色详情
- 时间线 (用户动态)
- 超展开 (帖吧)
- 时光机 (个人中心) / 用户空间
- 不限于搜索 / 排行榜 / 索引 / 目录 / 每日放送 / 年鉴 / 标签

## Update

v4.7.0 - 2021/02/18

—— 新增 ——

- [x] [发现] 推荐页面, 根据你的收藏, 猜测你喜欢的一些番剧 (实验性)
- [x] [发现] 2020 年鉴入口

—— 优化 ——

- [x] [首页收藏] 新增按放送时间排序,可在设置里面设定, 若设置还会显示放送星期几
- [x] [发现] 优化了所有年鉴的加载速度

—— 修复 ——

- [x] [角色] 适配同一角色多个声优的情况 (比如角色有中/日声优)
- [x] [每日放送] 修复部分条目星期错误, 因为数据是多源头的, 主要按日本时间为主, 会有部分番剧不准确, 仅供参考

—— 小圣杯 3.2.3 ——

- [x] [每周萌王] 显示平均拍价

[CHANGELOGS](https://github.com/czy0729/Bangumi/blob/master/CHANGELOG.MD)

## PowerBy

[[bangumi-subject]](https://github.com/czy0729/Bangumi-Subject) [![](https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi-Subject/badge)](https://www.jsdelivr.com/package/gh/czy0729/Bangumi-Subject) 条目快照 CDN

[[bangumi-oss]](https://github.com/czy0729/Bangumi-OSS) [![](https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi-OSS/badge)](https://www.jsdelivr.com/package/gh/czy0729/Bangumi-OSS) 热门条目封面和活跃用户头像对象存储 CDN

[[bangumi-rakuen]](https://github.com/czy0729/Bangumi-Rakuen) [![](https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi-Rakuen/badge)](https://www.jsdelivr.com/package/gh/czy0729/Bangumi-Rakuen) 帖子快照 CDN

[[bangumi-mono]](https://github.com/czy0729/Bangumi-Mono) [![](https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi-Mono/badge)](https://www.jsdelivr.com/package/gh/czy0729/Bangumi-Mono) 人物快照 CDN

[[bangumi-static]](https://github.com/czy0729/Bangumi-Static) [![](https://data.jsdelivr.com/v1/package/gh/czy0729/Bangumi-Static/badge)](https://www.jsdelivr.com/package/gh/czy0729/Bangumi-Static) 发现数据快照 CDN

[[bangumi-onair]](https://github.com/ekibun/bangumi_onair) [![](https://data.jsdelivr.com/v1/package/gh/ekibun/bangumi_onair/badge)](https://www.jsdelivr.com/package/gh/ekibun/bangumi_onair) [ekibun](https://github.com/ekibun) 的单集播放数据源

[[bangumi-mosaic-tile]](https://github.com/weizhenye/bangumi-mosaic-tile) [weizhenye](https://github.com/weizhenye) 的用户统计瓷砖库

[[bangumi-data]](https://github.com/bangumi-data/bangumi-data) 番组数据索引

[[bangumi-api]](https://github.com/bangumi/api) 官方接口

## Preview

#### 基础功能

<img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/1.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/1.png" width="100%" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/2.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/2.png" width="100%" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/3.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/3.png" width="100%" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/4.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/4.png" width="100%" />

#### 小圣杯 (角色模拟炒股) 模块

<img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/5.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/5.png" width="100%" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/6.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Static@master/data/screenshot/3.10.0/6.png" width="100%" />

(被墙看不见图的可以复制 alt 打开)

## 后续迭代

现版本基本用户需要的功能已齐全, 接下来会使用友盟收集到的埋点数据(匿名), 尝试学习数据分析, 根据结果计划后续版本迭代的方向.

[[bangumi-analysis]](https://github.com/czy0729/Bangumi-Analysis) 脱敏埋点数据

目前进程, 根据本 APP 最核心的页面 (条目), 先进行整合数据并进行简单的可视化模块搭建, 以下根据最近 30 天埋点数据情况, 显示平均一天事件触发次数和百分比

<img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/1.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/1.png" width="188" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/2.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/2.png" width="188" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/3.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/3.png" width="188" /><img src="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/4.png" alt="https://cdn.jsdelivr.net/gh/czy0729/Bangumi-Analysis@master/screenshots/4.png" width="188" />

可以一眼了解到哪些功能需求强, 可以进行迭代, 也能明确哪些功能需求偏低, 适当进行隐藏或者移除. 发现存在 0 数据的事件, 可以考虑是不是埋点写法出了问题.

## 免责声明

- 本项目所有数据信息均来自各种网站，不提供任何形式的媒体下载、直接播放和修改功能
- 本项目承诺不保存任何第三方用户信息
- 本项目代码仅供学习交流，不得用于商业用途，若侵权请联系

## 本地开发

** 提醒 **

最近项目使用的 Expo SDK 因官方问题, 现最新版本想在电脑上面跑起来需要手动修改包.
你可以 fork 完后切换到 tag v4.0.0 的版本进行尝试跑起来

** 下载 **

推荐使用 npm, 切换 npm 官方源可以按下面操作

```bash
npm config set registry https://registry.npmjs.org
```

国内环境, 首次下载包有可能会异常地慢

```bash
git clone https://github.com/czy0729/Bangumi
cd ./Bangumi
npm install // 不推荐使用 yarn, 因为后面会魔改 node_modules, 使用 yarn 会被识别并还原
```

** 前置操作, 否则 App 内图标均为问号 **

自定义 Iconfont, 把 assets/iconfont/ 下的目录 `vector-icons` 和 node_modules/@expo/ 下的目录 `vector-icons` 合并 (注意是合并不是替换, 假如不行, 就手动把文件夹下面的 2 个文件, 找到对应的地方替换, 观察文件修改日期是不是变化了), 原理其实就是用自己的字体文件把原有的 AntDesign 字体文件替换掉, 之后引用 Icon.AntDesign .

** 启动 **

```bash
npm start // [iOS] 之后点击在浏览器打开的Expo Developer Tools, 点击 Run on iOS simulator

// or

react-native run-android // [android] 真机 usb 调试
```

本人一直在 mac 环境下开发, iOS 跑 expo 客户端调试, 安卓跑真机 usb 调试. window 用法不详, 也不清楚到底能不能运行, 可以查阅 [Expo](https://github.com/expo/expo) 官方文档. 可以选择同一局域网真机扫码调试, 需先下载 Expo App.


CHANGELOG.MD
# CHANGELOG

v4.7.0 - 2021/02/18

—— 新增 ——

- [x] [发现] 推荐页面, 根据你的收藏, 猜测你喜欢的一些番剧 (实验性)
- [x] [发现] 2020 年鉴入口

—— 优化 ——

- [x] [首页收藏] 新增按放送时间排序,可在设置里面设定, 若设置还会显示放送星期几
- [x] [发现] 优化了所有年鉴的加载速度
- [x] [超展开] 帖子收藏会同步到云

—— 修复 ——

- [x] [角色] 适配同一角色多个声优的情况 (比如角色有中/日声优)
- [x] [每日放送] 修复部分条目星期错误, 因为数据是多源头的, 主要按日本时间为主, 会有部分番剧不准确, 仅供参考

—— 小圣杯 3.2.3 ——

- [x] [每周萌王] 显示平均拍价

v4.6.4 - 2021/01/30

—— 优化 ——

- [x] 优化了启动后的流畅度
- [x] [首页进度] 优化 APP 启动时渲染的条目数量, 加速滑动响应
- [x] [帖子] APP 内很多元素左右分布的 UI, 都会把元素调整到左边, 浏览更舒服
- [x] [私信] 打字的工具条增加图床入口
- [x] [索引] 添加方格布局, 和向前/后一个月的按钮
- [x] [排行榜] 切换筛选条件时列表不会再闪一下
- [x] 右上角 type UI 迭代

—— 修复 ——

- [x] [安卓] 修复了安卓设备黑暗模式, 页面切换的时候, 掉帧时右侧会出现白色的问题
- [x] [搜索] 修复查询人物失败的问题
- [x] [PM] UI 迭代, 与吐槽页一致
- [x] 游客访问敏感图片, 使用默认图片

v4.6.2 - 2021/01/18

—— 优化 ——

- [x] [条目] 单行本应使用系列 id 查询源头, 因为源头都是基于系列来搜索的
- [x] [条目] 小说|漫画系列, 若动画化, 顶部系列显示动画
- [x] [找漫画] 部分功能优化
- [x] 加速 Tabbar 点击响应

—— 修复 ——

- [x] 修复了启动页不是进度时, 进度页面无法点击的问题

—— 开发 ——

- [x] [HD] 灰度测试 HD 漫画源头
- [x] 加入阻止短时间内发送完全相同请求 HTML 的机制
- [x] 使用更 mini 版 bangumi-data, 缩小包体积, 1.2m -> 660k
- [x] 横向 ScrollView, 触发横向移动才渲染看不见的 Item

v4.6.0 - 2021/01/13

—— 新增 ——

- [x] [找番剧 | 文库 | 漫画]
  - [x] 新增找漫画
  - [x] 新增方格布局
  - [x] 类型标签支持长按多选
  - [x] 数据显示分数和排名, 排名优先度高于分数 (排名需要一定数量人打过分才会有)
  - [x] 更新了 1 月新番数据
  - [x] 一次最多显示前 500 条数据, 请活用筛选

—— 优化 ——

- [x] [条目] 优化源头搜索(说明一下: 源头跟本应用和作者都没任何关系, 部分源头广告多可以考虑不看或者装个带广告过滤的浏览器)
- [x] 方格样式番剧列表、分页块、筛选栏 UI 迭代
- [x] 强制设置不能分屏使用本应用 (分屏会卡死手机, 当前不知道怎么解决, 暂时屏蔽)

—— 修复 ——

- [x] [首页进度] 修复方格模式下, 章节数大于 24 格子显示不全的问
- [x] 修复再次点击底部 TabBar 图标到顶, 当页面没有数据会报错的问题

—— 开发 ——

- [x] 按需引入图标, 减少了 1M 打包体积

v4.5.0 - 2020/12/27

—— 新增 ——

- [x] [设置] 设置可上传到云端, 下载并恢复, 不用再怕重装后丢失

—— 优化 ——

- [x] [首页] 5 个底部 TabBar 图标, 再点击时会滚动到顶并刷新当前列表数据
- [x] [首页收藏] 条目增加直接查询播放源按钮 (需要在设置打开)
- [x] [条目] 未上映的番剧会单独显示上映日期
- [x] [条目] 页面内每个模块的标题点击都能折叠, 且通过分析埋点后, 使用率过低的模块默认折叠
- [x] [帖子] +1 和子楼层引用的 UI 迭代
- [x] [帖子] 楼层的引用回复默认最多显示 3 行, 点击展开
- [x] [帖子] 默认过滤掉用户已删除的楼层 (可在超展开设置中设置不过滤)
- [x] [时光机] 追加按 [网站评分] 排序
- [x] [找番剧] 条目分数使用更准确的最近快照
- [x] [找番剧] 删除无数据的标签
- [x] [设置] 隐藏不常用设置

—— 修复 ——

- [x] [个人设置] 不符合格式的数据会被修复, 有问题请再次保存即可
- [x] [吐槽 | 私信] 修复进入页面没有自动滚动到底的问题
- [x] [iOS] 修复 iphone12 系列 屏幕分辨率问题
- [x] [安卓] 修复用户曾用名错误行高导致昵称显示不全问题

v4.4.4 - 2020/12/14

—— 优化 ——

- [x] 更新各种静态数据
- [x] 修复部分已知问题
  - [x] 在线源头域名云同步机制, 修复播放源域名失效问题
- [x] [电波提醒] 提醒 tab, 完全相同文案的项会合并
- [x] [安卓] 所有页面点击顶部(刘海附近)可以回到顶部, 与 iOS 表现一致

v4.4.2 - 2020/11/30

—— 新增 ——

- [x] 提升好友在 App 内的表现, 不限于:
  - [x] 发现首页展示好友动态
  - [x] 会自动记录好友改名, 在部分常用页面一并显示曾用名

—— 优化 ——

- [x] 安卓键盘优化?
- [x] 收藏项的标签分割显示优化

—— 修复 ——

- [x] [帖子] 修复了帖子里楼层中自己的头像有时候依然是旧头像的问题
- [x] [帖子] 留言输入框, 安卓暂时屏蔽控制光标的逻辑

—— 小圣杯 3.2.0 ——

- [x] [我的持仓] 新增批量分享, 多选后自动把带链接的分享文案复制到粘贴板
- [x] [粘贴板] 识别剪贴板上比如 QQ 群分享带多角色链接的消息, 生成自定义列表 (页面支持一键参与 ico)
- [x] 交易和资产重组页面, 增加关联角色入口

v4.4.0 - 2020/11/19

—— 优化 ——

- [x] [帖子] 子楼层过多时自动折叠 (暂定显示 5 条子回复)
- [x] [目录详情] 列表支持排序
- [x] 吐槽页 UI 迭代

—— 小圣杯 3.1.0 ——

- [x] [我的委托] 增加按挂单数排序
- [x] [我的委托] 一键取消挂单和拍卖
- [x] [我的道具] 修复闪退问题
- [x] [我的道具] 虚空道标和混沌魔方没有成塔的角色不应出现在左边列表
- [x] [高级功能] 拍卖推荐可以按股份排序
- [x] 所有角色在涉及资产变动操作后, 会主动联动我的持仓, 减少了数据重新加载, 增强持股数更新速度

v4.3.0 - 2020/11/07

—— 新增 ——

- [x] [条目] 包含本条目的目录进入更多列表页

—— 优化 ——

- [x] [目录] 列表项 UI 迭代
- [x] [目录详情] 列表项没有评分数据, 使用了评分快照

—— 修复 ——

- [x] 修复了有些地方头像点击没反应的问题
- [x] 修复了找番剧数据不显示的问题

—— 小圣杯 3.0.0 ——

- [x] [我的持仓] 批量操作增加批量出售(出售给英灵殿), 批量挂卖单(按当前价挂卖单)
- [x] [我的持仓] 批量模式下支持多选(每次点击从最后一个选择的位置多选 10 个)
- [x] [资产重组] 新增消耗最低数量股数测试献祭效率 (效率 = 返还资金 / 当前总价值 的百分比)
- [x] [我的道具] 列表增加 lv 筛选
- [x] [我的道具] 星光碎片支持低级活股补充高级圣殿
- [x] 部分列表 Tabs 显示列表角色数量
- [x] 金额缩略模式下, 全局的金额数字都会缩略
- [x] 优化列表等级筛选下拉菜单
- [x] 搜索强化
- [x] 角色界面现在可以启动 ICO

v4.2.2 - 2020/10/29

—— 新增 ——

- [x] [条目] 包含本条目的目录模块 (更多的列表页面开发中)

—— 优化 ——

- [x] [排行榜] 切换年月份重置到第一页
- [x] [吐槽] 用户头像加载应倒序加载

—— 修复 ——

- [x] 修复了在首页登出或者登陆状态过期后, 没出现进入登陆页面按钮的问题

—— 小圣杯 2.8.0 ——

- [x] [番市首富] 展示数据更合理
- [x] 列表项显示人物所属条目, 右侧菜单添加进入关联人物列表
- [x] 条目关联人物列表
- [x] 用户头像角标显示用户在线状态 (实验性, 1 天内为绿色, 3 天内为橙色, 其余不显示)
- [x] ico 列表当 lv0 人数 10-14 人, 高亮显示

v4.2.0 - 2020/10/24

—— 新功能 ——

- [x] [超展开] 通过搜索引擎模糊搜索帖子(实验性, 非高级会员有限制)
- [x] [空间] 可以查询到用户历史帖子 (非高级会员有限制)

—— 优化 ——

- [x] [帖子] 吐槽右侧筛选按钮 UI 优化

—— 修复 ——

- [x] [设置] 主题颜色跟随系统优化, 只支持 android 10 及以上
- [x] [个人设置] 修复了输入框难操作的问题

—— 小圣杯 ——

- [x] 刮刮乐弹窗后在弹窗内增加再刮一次的按钮, 并显示下次价格
- [x] 列表增加等级筛选
- [x] 修复白天模式刮刮乐弹窗背景颜色错误问题
- [x] 修复当前版本所有失效的功能
- [x] 修复无法获取当前拍卖人数的问题
- [x] 修复圣殿图太小图片没占满的问题

v4.1.0 - 2020/10/18

—— 新功能 ——

- [x] [条目] 章节一览页面
- [x] [条目] 可能的话会搜索动画预览图
- [x] [设置] 添加了一个网络探针工具页面, 用于检测 App 资源域访问情况

—— 优化 ——

- [x] [条目] 优化了章节输入框后面/??的数字, 会准确匹配已确定集数
- [x] [条目] 优化了收藏盒下全站收藏人数文字, 并可以跳转到用户评分页面对应 tabs
- [x] [条目] 文案修改: 评论 -> 日志, 讨论版 -> 帖子
- [x] [条目] 加强了前传和续集关系 UI

—— 小圣杯 ——

- [x] 修复了人物数据接口被删除导致不显示的问题

v4.0.1 - 2020/10/10

—— 修复 ——

- [x] [进度] 修复了长篇动画例如海贼王, 操作看过后看过的集数不正确的问题
- [x] [条目] 修复了漫画的角色更多列表无法加载的问题
- [x] [条目] 修复了制作人员列表进入制作人员详情无法加载数据的问题
- [x] [帖子] 修复了某些帖子报错的问题
- [x] 修复了部分图和和浏览大图无法加载的问题
- [x] 修复了部分机型启动闪退的问题

v4.0.0 - 2020/10/06

—— 新增 ——

- [x] [首页进度] 可以在设置里面把在玩的游戏也显示 (默认关闭)
- [x] [条目] 吐槽可以根据评分筛选浏览
- [x] [条目] 现在单行本条目, 会有本系列的关联 UI 表现
- [x] [设置] 现在可以选择启动 APP 时是否跟随系统主题
- [x] 条目封面拟物, 增加辨识度 (默认开启, 设置可关闭, 书籍形状、音乐碟片形状、游戏卡带形状)

—— 优化 ——

- [x] [条目] 标题旁显示更细的条目类型 (动画、漫画系列、漫画、小说系列等)
- [x] 升级底层框架版本 (RN@0.59 -> 0.62), 默认修复部分现有问题
- [x] 因框架新版本, android/pageViewer 废弃, 所有旧 Tabs 替换新库实现
- [x] 找番剧、文库的静态数据迁移云端, 缩小包体积
- [x] 列表里自己的头像不走快照, 保证最新
- [x] 屏蔽柠萌瞬间相关, 下版本完全移除
- [x] 现在不会重复进入参数完全相同的页面
- [x] 请求中的转圈可以点击关闭, 避免请求超时无法操作
- [x] 优化了漫画阅读器的图片加载顺序

—— 修复 ——

- [x] [条目] 修复页面进入时, 数据加载慢会重复多次请求吐槽的问题
- [x] [时间胶囊] 修复了范围为自己时, 看过 x of x 集形式文字丢失问题
- [x] 修复漫画无法搜索源头的问题 , 增加了提示
- [x] 修复日志详情页面标题为 undefined 的问题

v3.12.0 -2020/09/13

—— 新增 ——

- [x] [个人设置] 支持修改昵称、签名, 可自定义背景、头像 (非网站头像, 网站头像暂时无法修改, 修改的仅为 APP 内和 APP 用户查看空间时显示)
- [x] [安卓] Deep Link, 外部 bgm 链接可以打开 APP 对应页面

—— 优化 ——

- [x] 增加了操作类请求后轻震动反馈 (不需要可到设置关闭)

—— 修复 ——

- [x] [时间胶囊] 修复了范围为自己时, 看过 x of x 集形式文字丢失问题

v3.11.0 - 2020/09/03

—— 新增 ——

- [x] [发现] 找文库 (有记录的都是有源头的)
- [x] [条目] 部分热门轻小说加入源头

—— 优化 ——

- [x] [找番剧] 更新数据, 新增国漫数据 (当前收录 2381 个条目)
- [x] [时间线] 多图片横向滚动右边间距优化

—— 修复 ——

- [x] [条目] 修复了评论和讨论版不显示的问题

v3.10.0 - 2020/07/30

—— 新增 ——

- [x] [条目] 好友评分详细列表
- [x] [时光机] 新增时间线页面

—— 优化 ——


App.v2.js
/*
 * @Author: czy0729
 * @Date: 2019-03-30 19:25:19
 * @Last Modified by: czy0729
 * @Last Modified time: 2019-10-21 09:47:00
 */
import React from 'react'
import { StyleSheet, View, NativeModules } from 'react-native'
import { useScreens } from 'react-native-screens'
import { AppLoading } from 'expo'
import * as Font from 'expo-font'
import { Provider } from '@ant-design/react-native'
import { ImageViewer } from '@components'
import Stores, { systemStore } from '@stores'
import { observer } from '@utils/decorators'
import { hm } from '@utils/fetch'
import _ from '@styles'
import theme from '@styles/theme'
import Navigations from './navigations/index'

/**
 * https://reactnavigation.org/docs/zh-Hans/react-native-screens.html
 */
useScreens()

console.disableYellowBox = true

/**
 * 能打印循环引用
 */
global.log = (value, space) => {
  const handleCircular = () => {
    const cache = []
    const keyCache = []
    return (key, value) => {
      if (typeof value === 'object' && value !== null) {
        const index = cache.indexOf(value)
        if (index !== -1) {
          return `[Circular ${keyCache[index]}]`
        }
        cache.push(value)
        keyCache.push(key || 'root')
      }
      return value
    }
  }
  console.log(JSON.stringify(value, handleCircular(), space))
}

/**
 * 马上隐藏SplashScreen
 */
const { ExponentSplashScreen: SplashScreen = {} } = NativeModules
export function preventAutoHide() {
  if (SplashScreen.preventAutoHide) {
    SplashScreen.preventAutoHide()
  }
}
export function hide() {
  if (SplashScreen.hide) {
    SplashScreen.hide()
  }
}

export default
@observer
class App extends React.Component {
  state = {
    isLoadingComplete: false
  }

  componentDidMount() {
    this.loadResourcesAsync()
  }

  componentDidCatch(error) {
    hm(`error?error=${error}`, '错误')
  }

  loadResourcesAsync = async () => {
    const res = Promise.all([
      Stores.init()
      // Asset.loadAsync([]),
      // Font.loadAsync({
      //   bgm: require('./assets/fonts/AppleColorEmoji.ttf')
      // })
    ])
    await res

    this.handleFinishLoading()
    Promise.all([
      Font.loadAsync({
        bgm: require('./assets/fonts/AppleColorEmoji.ttf')
      })
    ])

    return res
  }

  handleLoadingError = error => {
    console.warn(error)
  }

  handleFinishLoading = () => {
    this.setState(
      {
        isLoadingComplete: true
      },
      () => {
        // hide()
      }
    )
  }

  closeImageViewer = () => {
    systemStore.closeImageViewer()
  }

  render() {
    // const { skipLoadingScreen } = this.props
    const { isLoadingComplete } = this.state
    if (!isLoadingComplete) {
      return null
    }

    // if (!isLoadingComplete && !skipLoadingScreen) {
    //   return (
    //     <AppLoading
    //       startAsync={this.loadResourcesAsync}
    //       onFinish={this.handleFinishLoading}
    //       onError={this.handleLoadingError}
    //     />
    //   )
    // }

    const { visible, imageUrls } = systemStore.imageViewer
    return (
      <View style={styles.container}>
        <Provider theme={theme}>
          <Navigations />
        </Provider>
        <ImageViewer
          visible={visible}
          imageUrls={imageUrls}
          onCancel={this.closeImageViewer}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: _.colorBg
  }
})
